### PR TITLE
osd/OSDMap.cc: add more info in json output of osd stat

### PR DESCRIPTION
**Pull request by shenhang**
Fixes: https://tracker.ceph.com/issues/41718

Signed-off-by: Shen Hang <harryshen18@gmail.com>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`

</details>

### CODE REVIEW

File: src/osd/OSDMap.cc [hunk 1]
```diff
@@ -3968,7 +3968,13 @@ void OSDMap::print_summary(Formatter *f, ostream& out,
     f->dump_int("epoch", get_epoch());
     f->dump_int("num_osds", get_num_osds());
     f->dump_int("num_up_osds", get_num_up_osds());
+    if (last_up_change != utime_t()) {
+      f->dump_string("osd up since", utimespan_str(now - last_up_change));
```
Comment (socketpair) [inline src/osd/OSDMap.cc]: Please replace spaces with underscores in field names.
Use dump_int + a number.

File: src/osd/OSDMap.cc [hunk 2]
```diff
@@ -3968,7 +3968,13 @@ void OSDMap::print_summary(Formatter *f, ostream& out,
     f->dump_int("epoch", get_epoch());
     f->dump_int("num_osds", get_num_osds());
     f->dump_int("num_up_osds", get_num_up_osds());
+    if (last_up_change != utime_t()) {
```
Comment (socketpair) [inline src/osd/OSDMap.cc]: Remove this check. What is purpose of it ?

**Comment by neha-ojha** [comment]
retest this please

File: src/osd/OSDMap.cc [hunk 3]
```diff
@@ -3965,10 +3965,13 @@ void OSDMap::print_summary(Formatter *f, ostream& out,
 			   const string& prefix, bool extra) const
 {
   if (f) {
+    utime_t now = ceph_clock_now();
     f->dump_int("epoch", get_epoch());
     f->dump_int("num_osds", get_num_osds());
     f->dump_int("num_up_osds", get_num_up_osds());
+    f->dump_int("osd_up_since", (now - last_up_change).to_msec() / 1000ull);
```
Comment (socketpair) [inline src/osd/OSDMap.cc]: `ull` is useless.

Comment (shenhang) [reply src/osd/OSDMap.cc]: > `ull` is useless.

ok,changed

**Comment by socketpair** [comment]
LGTM!

**Comment by shenhang** [comment]
> nit: can you use "osd/OSDMap.cc:" as the prefix for your commit title instead of "rados:"
> See https://github.com/ceph/ceph/blob/master/SubmittingPatches.rst#3-describe-your-changes for more details.

changed.Thanks

File: src/osd/OSDMap.cc [hunk 4]
```diff
@@ -3965,10 +3965,13 @@ void OSDMap::print_summary(Formatter *f, ostream& out,
 			   const string& prefix, bool extra) const
 {
   if (f) {
+    utime_t now = ceph_clock_now();
     f->dump_int("epoch", get_epoch());
     f->dump_int("num_osds", get_num_osds());
     f->dump_int("num_up_osds", get_num_up_osds());
+    f->dump_int("osd_up_since", (now - last_up_change).to_msec() / 1000);
```
Comment (dzafman) [inline src/osd/OSDMap.cc]: Is output of some large number of seconds really that useful?

Comment (shenhang) [reply src/osd/OSDMap.cc]: > Is output of some large number of seconds really that useful?

@socketpair I also think the number is very large. Could I change it back to utimespan_str?

Comment (socketpair) [reply src/osd/OSDMap.cc]: nope, JSON is a machine-centric output. number of seconds is the most convenient (useful) information. Human readable datetime requires parsing.

File: src/osd/OSDMap.cc [hunk 5]
```diff
@@ -3966,10 +3966,15 @@ void OSDMap::print_summary(Formatter *f, ostream& out,
...
```
Comment (tchaikov) [inline src/osd/OSDMap.cc]: i'd prefer printing the information conveyed by osdmap itself which does not change overtime.

Comment (socketpair) [reply src/osd/OSDMap.cc]: Personally, I dont' understand what you wanted to say...is it bad or good, or what is wrong...

Comment (tchaikov) [reply src/osd/OSDMap.cc]: @socketpair i think we should just print the timestamp of `last_up_change`.

Comment (socketpair) [reply src/osd/OSDMap.cc]: Yep, you are right. Agree.

Comment (shenhang) [reply src/osd/OSDMap.cc]: ok,changed.

Comment (shenhang) [reply src/osd/OSDMap.cc]: @tchaikov This pr last a little long ,pls have a look here.

File: src/osd/OSDMap.cc [hunk 6]
```diff
@@ -3966,10 +3966,15 @@ void OSDMap::print_summary(Formatter *f, ostream& out,
...
 {
   if (f) {
+    utime_t now = ceph_clock_now();
     f->dump_int("epoch", get_epoch());
     f->dump_int("num_osds", get_num_osds());
     f->dump_int("num_up_osds", get_num_up_osds());
+    f->dump_int("osd_up_since", last_up_change.to_msec() / 1000);
     f->dump_int("num_in_osds", get_num_in_osds());
+    f->dump_int("osd_in_since", last_in_change.to_msec() / 1000);
+    f->dump_bool("full", test_flag(CEPH_OSDMAP_FULL) ? true : false);
```
Comment (tchaikov) [inline src/osd/OSDMap.cc]: s/ ? true : false//

File: src/osd/OSDMap.cc [hunk 7]
```diff
@@ -3966,10 +3966,15 @@ void OSDMap::print_summary(Formatter *f, ostream& out,
 			   const string& prefix, bool extra) const
 {
   if (f) {
+    utime_t now = ceph_clock_now();
```
Comment (tchaikov) [inline src/osd/OSDMap.cc]: please drop this line.

File: src/osd/OSDMap.cc [hunk 8]
```diff
@@ -3966,10 +3966,15 @@ void OSDMap::print_summary(Formatter *f, ostream& out,
...
     f->dump_int("epoch", get_epoch());
     f->dump_int("num_osds", get_num_osds());
     f->dump_int("num_up_osds", get_num_up_osds());
+    f->dump_int("osd_up_since", last_up_change.to_msec() / 1000);
     f->dump_int("num_in_osds", get_num_in_osds());
+    f->dump_int("osd_in_since", last_in_change.to_msec() / 1000);
+    f->dump_bool("full", test_flag(CEPH_OSDMAP_FULL) ? true : false);
+    f->dump_bool("nearfull", test_flag(CEPH_OSDMAP_NEARFULL) ? true : false);
```
Comment (tchaikov) [inline src/osd/OSDMap.cc]: ditto.

File: src/osd/OSDMap.cc (hunk 6 above)
Comment (shenhang) [reply src/osd/OSDMap.cc]: > s/ ? true : false//

done

File: src/osd/OSDMap.cc (hunk 7 above)
Comment (shenhang) [reply src/osd/OSDMap.cc]: ok done

File: src/osd/OSDMap.cc (hunk 8 above)
Comment (shenhang) [reply src/osd/OSDMap.cc]: done

**Comment by tchaikov** [comment]
- http://pulpito.ceph.com/kchai-2019-12-21_03:01:01-rados-wip-kefu-testing-2019-12-21-0133-distro-basic-mira/
- http://pulpito.ceph.com/kchai-2019-12-23_16:20:31-rados-wip-kefu-testing-2019-12-21-0133-distro-basic-smithi/

failures tracked by

- https://tracker.ceph.com/issues/43396
- https://tracker.ceph.com/issues/43415
- https://tracker.ceph.com/issues/43414
- https://tracker.ceph.com/issues/43416
- https://tracker.ceph.com/issues/43394

**Pull request merged**